Remotion/Data/DomainObjects.UnitTests/Persistence/ReadOnlyStorageProviderDecoratorTest.cs

```diff
@@ -0,0 +1,131 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using Moq;
+using NUnit.Framework;
+using Remotion.Data.DomainObjects.Persistence;
+using Remotion.Data.DomainObjects.Queries;
+using Remotion.Data.DomainObjects.Queries.Configuration;
+using Remotion.Development.NUnit.UnitTesting;
+
+namespace Remotion.Data.DomainObjects.UnitTests.Persistence
+{
+  public class ReadOnlyStorageProviderDecoratorTest
+  {
+    [Test]
+    [TestCase(QueryType.CollectionReadOnly)]
+    [TestCase(QueryType.CustomReadOnly)]
+    [TestCase(QueryType.ScalarReadOnly)]
+    public void ExecuteCollectionQuery_WithReadOnlyQueryType_CallsInnerStorageProvider (QueryType queryType)
+    {
+      var queryStub = new Mock<IQuery>();
+      queryStub.Setup(_ => _.QueryType).Returns(queryType);
+      var innerReadOnlyStorageProviderStub = new Mock<IReadOnlyStorageProvider>();
+      var readOnlyStorageProviderDecorator = new ReadOnlyStorageProviderDecorator(innerReadOnlyStorageProviderStub.Object);
+
+      Assert.That(() => readOnlyStorageProviderDecorator.ExecuteCollectionQuery(queryStub.Object), Throws.Nothing);
+
+      innerReadOnlyStorageProviderStub.Verify(_ => _.ExecuteCollectionQuery(queryStub.Object), Times.Once);
+    }
+
+    [Test]
+    [TestCase(QueryType.CollectionReadOnly)]
+    [TestCase(QueryType.CustomReadOnly)]
+    [TestCase(QueryType.ScalarReadOnly)]
+    public void ExecuteCustomQuery_WithReadOnlyQueryType_CallsInnerStorageProvider (QueryType queryType)
+    {
+      var queryStub = new Mock<IQuery>();
+      queryStub.Setup(_ => _.QueryType).Returns(queryType);
+      var innerReadOnlyStorageProviderStub = new Mock<IReadOnlyStorageProvider>();
+      var readOnlyStorageProviderDecorator = new ReadOnlyStorageProviderDecorator(innerReadOnlyStorageProviderStub.Object);
+
+      Assert.That(() => readOnlyStorageProviderDecorator.ExecuteCustomQuery(queryStub.Object), Throws.Nothing);
+
+      innerReadOnlyStorageProviderStub.Verify(_ => _.ExecuteCustomQuery(queryStub.Object), Times.Once);
+    }
+
+    [Test]
+    [TestCase(QueryType.CollectionReadOnly)]
+    [TestCase(QueryType.CustomReadOnly)]
+    [TestCase(QueryType.ScalarReadOnly)]
+    public void ExecuteScalarQuery_WithReadOnlyQueryType_CallsInnerStorageProvider (QueryType queryType)
+    {
+      var queryStub = new Mock<IQuery>();
+      queryStub.Setup(_ => _.QueryType).Returns(queryType);
+      var innerReadOnlyStorageProviderStub = new Mock<IReadOnlyStorageProvider>();
+      var readOnlyStorageProviderDecorator = new ReadOnlyStorageProviderDecorator(innerReadOnlyStorageProviderStub.Object);
+
+      Assert.That(() => readOnlyStorageProviderDecorator.ExecuteScalarQuery(queryStub.Object), Throws.Nothing);
+
+      innerReadOnlyStorageProviderStub.Verify(_ => _.ExecuteScalarQuery(queryStub.Object), Times.Once);
```

Comment:
even though Moq supports this, calling Verify on a test double that's named 'stub' is semantically incorrect and will lead to false assumptions about test correctness in the future. 

In general, I recommend supplying a fake return value and asserting that for the returnvalue instead of asserting the actual call whenever this is possible. Another benefit of this is that right now, the test actually does not assert that the correct value is returned. You could have the ExecuteSqualarQuery method on the decorator return NULL and it would not fail the test.

Remidner: affects all three tests, please.

Reply:
I would keep the Verify setup for the collection tests because the setup with datacontainers and object ids is too convoluted for this otherwise very simple test.